src/caffe/layers/batch_norm_layer.cpp

```diff
@@ -0,0 +1,230 @@
+#include <algorithm>
+#include <vector>
+
+#include "caffe/common_layers.hpp"
+#include "caffe/layer.hpp"
+#include "caffe/util/math_functions.hpp"
+
+namespace caffe {
+
+template <typename Dtype>
+void BatchNormLayer<Dtype>::LayerSetUp(const vector<Blob<Dtype>*>& bottom,
+      const vector<Blob<Dtype>*>& top) {
+  BatchNormParameter param = this->layer_param_.batch_norm_param();
```

Comment:
The tests pass, so it's unlikely to be mathematically wrong.  Note that I stripped out the shift and scaling factors, which is probably the main difference.  You can try adding those back in using DummyData/Conv/Eltwise operations.  In fact, it may be a good to have an example of this.